benchmark/httpapi.py

```diff
@@ -0,0 +1,218 @@
+        Return matching results.
+        """
+        matching = [
+            r for r in self._results.viewvalues()
+            if filter.viewitems() <= r.viewitems()
+        ]
+        return succeed(matching)
+
+    def delete(self, id):
+        """
+        Delete a result by the given identifier.
+        """
+        try:
+            del self._results[id]
+            return succeed(None)
+        except KeyError:
+            return fail(ResultNotFound())
+
+
+class BenchmarkAPI_V1(object):
+    """
+    API for storing and accessing benchmarking results.
+
+    :ivar IBackend backend: The backend for storing the results.
+    """
+    app = Klein()
+    version = 1
+
+    def __init__(self, backend):
+        """
+        :param IBackend backend: The backend for storing the results.
+        """
+        self.backend = backend
+
+    @app.handle_errors(ResultNotFound)
+    def _not_found(self, request, failure):
+        request.setResponseCode(NOT_FOUND)
+        return ""
+
+    @app.route("/benchmark-results", methods=['POST'])
+    def post(self, request):
+        """
+        Post a new benchmarking result.
+
+        :param twisted.web.http.Request request: The request.
+        """
+        request.setHeader(b'content-type', b'application/json')
+        try:
+            json = loads(request.content.read())
+        except ValueError as e:
+            err(e, "failed to parse result")
+            request.setResponseCode(BAD_REQUEST)
+            return dumps({"message": e.message})
+
+        d = self.backend.store(json)
+
+        def stored(id):
+            msg("stored result with id {}".format(id))
+            result = {"version": self.version, "id": id}
+            response = dumps(result)
+            location = urljoin(request.path + '/', id)
+            request.setHeader(b'Location', location)
+            request.setResponseCode(CREATED)
+            return response
+
+        d.addCallback(stored)
+        return d
+
+    @app.route("/benchmark-results/<string:id>", methods=['GET'])
+    def get(self, request, id):
+        """
+        Get a previously stored benchmarking result by its ID.
+
+        :param twisted.web.http.Request request: The request.
+        :param str id: The identifier.
+        """
+        request.setHeader(b'content-type', b'application/json')
+        d = self.backend.retrieve(id)
+
+        def retrieved(result):
+            response = dumps(result)
+            return response
+
+        d.addCallback(retrieved)
+        return d
+
+    @app.route("/benchmark-results/<string:id>", methods=['DELETE'])
+    def delete(self, request, id):
+        """
+        Delete a previously stored benchmarking result by its ID.
+
+        :param twisted.web.http.Request request: The request.
+        :param str id: The identifier.
+        """
+        request.setHeader(b'content-type', b'application/json')
+        request.setResponseCode(NO_CONTENT)
+        return self.backend.delete(id)
```

Comment:
This should be handled by `_not_found` method defined above. It converts `ResultNotFound` exception to a `404` reponse.